### ISSUE TITLE
only iterate over unique subset of icols

### ISSUE THREAD

**Issue by ben-schwen** [Rdatatable/data.table#6630]
On first read I think "this can be doing a lot of duplicate work since it's basically `for (icol in icols) icol==icols`", but actually it only applies to the subset `icols[icols != xcols & i_merge_type %in% c("integer", "double", "complex")]`.
_Originally posted by @MichaelChirico in https://github.com/Rdatatable/data.table/pull/6603#discussion_r1861329020_

**Comment by TusharNaugain** [comment]
> ```
>           On first read I think "this can be doing a lot of duplicate work since it's basically `for (icol in icols) icol==icols`", but actually it only applies to the subset `icols[icols != xcols & i_merge_type %in% c("integer", "double", "complex")]`.
> ```
> 
> _Originally posted by @MichaelChirico in [#6603 (comment)](https://github.com/Rdatatable/data.table/pull/6603#discussion_r1861329020)_
> 
> minimal example
> 
> ```r
> x = data.table(a=1L)
> y = data.table(c=1L, d=1)
> y[x, on=.(c == a, d == a)]
> ```
> 
> In `bmerge` we iterate over `icols` with `for (a in seq_along(icols)`, hence in this case we epxlore twice `a=1` while once would be enough.
> 
> Corresponding bmerge line:
> 
> https://github.com/Rdatatable/data.table/blob/546259ddaba0e8ab1506729113688f85ca2986fd/R/bmerge.R#L37

Hey there,

I noticed that in the current implementation, we iterate over icols which might cause redundant processing. To optimize this, we can create a unique subset of icols that excludes elements from xcols and matches the required i_merge_type. This way, each element is processed only once.

Here's a code snippet to achieve this

const icols = [/* your array of columns */];
const xcols = [/* your array of columns to exclude */];
const i_merge_type = [/* your merge type */];
const uniqueIcols = icols.filter(icol => !xcols.includes(icol) && i_merge_type.includes(icol));
for (const a in uniqueIcols) {
  
}
By using filter, we ensure that the iteration is done over a unique subset, thus avoiding duplicate work. This should help streamline the processing in bmerge.

**Comment by Abhishek2634** [comment]
Hi @ben-schwen  can we use this line:  if (nrow(i)) for (a in seq_along(unique(icols)))
if its okay,  can raise a PR for this ?

**Comment by MichaelChirico** [comment]
I don't think it will be so easy -- `icols` and `xcols` are paired vectors, and that loop assumes we can pull `icols[a]` and `xcols[a]`. Something like `cbind(icols, xcols)[!duplicated(icols), ]` also won't work since it might drop some `xcols` that need to be checked.

**Comment by jangorecki** [comment]
This loop does not work on the actual data but rather on the metadata, type matching, coerce. Therefore potential gains in performance may be non existing (close to zero). Before working on this one maybe good to see if that loops takes any time to compute.

**Comment by Suman1108-beep** [comment]
Hi, I’d like to look into this issue.
I’ll start by profiling this bmerge path on cases where icols contains duplicates to see whether the loop is a measurable bottleneck, and then report back with benchmarks before proposing any code changes.
